Client/Frontend/Sync/BraveCore/Bookmarks/BookmarkAPIExtensions.swift

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+    
+    func add(url: URL, title: String?, parentFolder: BookmarkNode? = nil) {
+        if let parentFolder = parentFolder {
+            createBookmark(withParent: parentFolder, title: title ?? "", with: url)
+        } else {
+            createBookmark(withTitle: title ?? "", url: url)
+        }
+    }
+    
+    func frc(parent: BookmarkNode?) -> BookmarksV2FetchResultsController? {
+        return Bookmarkv2Fetcher(parent, api: self)
+    }
+    
+    func foldersFrc(excludedFolder: BookmarkNode? = nil) -> BookmarksV2FetchResultsController? {
+        return Bookmarkv2ExclusiveFetcher(excludedFolder, api: self)
+    }
+    
+    func getChildren(forFolder folder: BookmarkNode, includeFolders: Bool) -> [BookmarkNode]? {
+        let result = folder.children.map({ $0 })
+        return includeFolders ? result : result.filter({ $0.isFolder == false })
+    }
+    
+    func byFrequency(query: String? = nil, completion: @escaping ([BookmarkNode]) -> Void) {
+        // Invalid query.. BraveCore doesn't store bookmarks based on last visited.
+        // Any last visited bookmarks would show up in `History` anyway.
+        // BraveCore automatically sorts them by date as well.
+        guard let query = query, !query.isEmpty else {
+            completion([])
+            return
+        }
+        
+        return search(withQuery: query, maxCount: 200, completion: { nodes in
+            completion(nodes.compactMap({ return !$0.isFolder ? $0 : nil }))
+        })
+    }
+    
+    func reorderBookmarks(frc: BookmarksV2FetchResultsController?, sourceIndexPath: IndexPath, destinationIndexPath: IndexPath) {
+        guard let frc = frc else {
+            return
+        }
+        
+        if let node = frc.object(at: sourceIndexPath),
+           let parent = node.parent ?? mobileNode {
+            
+            // Moving to the very last index.. same as appending..
+            if destinationIndexPath.row == parent.children.count - 1 {
+                node.move(toParent: parent)
+            } else {
+                node.move(toParent: parent, index: UInt(destinationIndexPath.row))
+            }
+            
+            // Notify the delegate that items did move..
+            // This is already done automatically in `Bookmarkv2Fetcher` listener.
+            // However, the Brave-Core delegate is being called before the move is actually complete OR too quickly
+            // So to fix it, we reload here AFTER the move is done so the UI can update accordingly.
+            frc.delegate?.controllerDidReloadContents(frc)
+        }
+    }
+    
+    func delete(_ bookmarkItem: BookmarkNode) {
+        if !bookmarkItem.isPermanentNode {
+            removeBookmark(bookmarkItem)
+        }
+    }
+    
+    func updateWithNewLocation(_ bookmarkItem: BookmarkNode, customTitle: String?, url: URL?, location: BookmarkNode?) {
+        if let location = location ?? mobileNode {
```

Comment:
Guard + log + return early instead of if-let for entire block of code

Reply:
Fix at https://github.com/brave/brave-ios/pull/4024/commits/72abd0fe27a683742ddce5ce85e361cb6d7dcfb2